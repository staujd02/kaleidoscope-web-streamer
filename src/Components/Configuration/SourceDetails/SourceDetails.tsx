import React from 'react';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Input from '@material-ui/core/Input';
import Switch from '@material-ui/core/Switch';
import TextField from '@material-ui/core/TextField';
import InputLabel from '@material-ui/core/InputLabel';
import FormGroup from '@material-ui/core/FormGroup';
import Grid from '@material-ui/core/Grid';

export default class SourceDetails extends React.Component<SourceDetailsProps, SourceDetailsState> {

    constructor(props: SourceDetailsProps) {
        super(props);
        this.state = {
            error: '',
        }
    }
    
    handleToggle = (name: string) => (event: any) => {
        const { checked } = event.target;
        this.handleErrors(() =>
            this.props.handleUpdate({
                ...this.props.source!,
                [name]: !checked
            }), name);
    }

    handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value } = event.target;
        this.handleErrors(() =>
            this.props.handleUpdate({
                ...this.props.source!,
                [name]: value
            }), name);
    }

    handleErrors = (call: () => void, optionalErrorState?: string) => {
        try {
            this.setState({ error: '' }, call);
        } catch (ex) {
            this.setState({
                error: optionalErrorState ? optionalErrorState : ex
            });
        }
    }

    render() {
        const { error } = this.state;
        return (
            <Card className="card source-detail">
                <CardContent>
                    <Typography className="title" color="textSecondary" gutterBottom>
                        Stream Details
                    </Typography>
                    {this.props.source === null &&
                        <Typography className="placeholder">
                            Select a stream to view it's details.
                        </Typography>}
                    {this.props.source !== null && (
                        <FormGroup>
                            <TextField name='title' label="Source Name" type="text" onChange={this.handleChange} value={this.props.source.title} />
                            <TextField name='source' className="beefy" label="URL" onChange={this.handleChange} type="url" helperText="Must be embeddable" value={this.props.source.source} />
                            <Grid>
                                <InputLabel id="durationLabel" error={error === 'duration'} className="inline-lb">Duration</InputLabel>
                                <Input id="duration" name="duration" onChange={this.handleChange} type="number" value={this.props.source.duration} />
                            </Grid>
                            <Grid>
                                <InputLabel className="inline-lb">Enabled</InputLabel>
                                <Switch id="enable-toggle" name='isEnabled' onClick={this.handleToggle('isEnabled')} type="button" checked={this.props.source.isEnabled} />
                            </Grid>
                        </FormGroup>
                    )}
                </CardContent>
            </Card>
        );
    }
}