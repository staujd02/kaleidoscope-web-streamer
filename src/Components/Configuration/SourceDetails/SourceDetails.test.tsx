import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { formatHTML } from '../../../TestUtilities/htmlFormatter';
import SourceDetails from './SourceDetails';

describe('The Source Detail Component', () => {

    let wrapper: ShallowWrapper<SourceDetailsProps, {}, SourceDetails>
    let props: SourceDetailsProps;

    let handleUpdate: jest.Mock;

    beforeEach(() => {
        props = {
            handleUpdate: (handleUpdate = jest.fn()),
            source: {
                duration: 6000,
                isEnabled: true,
                sortOrder: 5,
                source: "www.somewhere",
                title: "The Best Source",
                key: 3
            }
        }
        wrapper = shallow(<SourceDetails {...props} />);
    });

    it('renders correctly', () => expect(formatHTML(wrapper.html())).toMatchSnapshot());
    
    describe('given the user wants to make a change to a toggle button', () => {

        beforeEach(() => {
            wrapper.find('#enable-toggle').simulate('click', {
                target: {
                    checked: true,
                    name: wrapper.find('#enable-toggle').prop('name')
                }
            });
        });

        it("saved through the parent's callback", () => {
            expect(handleUpdate).toHaveBeenCalledWith({
                ...props.source,
                isEnabled: false
            });
        });
    });

    describe('given the user makes a change', () => {

        beforeEach(() => {
            wrapper.find('#duration').simulate('change', {
                target: {
                    value: "5000",
                    name: wrapper.find('#duration').prop('name')
                }
            });
        });

        it("saved through the parent's callback", () => {
            expect(handleUpdate).toHaveBeenCalledWith({
                ...props.source,
                duration: "5000"
            });
        });
    });

    describe('when the user inputs a bad value', () => {
        describe('given it throws an error', () => {
            const exceptionMessage = 'Blarg!';

            beforeEach(() => {
                handleUpdate.mockImplementation(() => {
                    throw new Error(exceptionMessage);
                });
                wrapper.find('#duration').simulate('change', {
                    target: {
                        value: "-5000",
                        name: "duration"
                    }
                });
            });

            it('shows the user what went wrong', () => {
                expect(wrapper.find('#durationLabel').prop('error')).toEqual(true);
            });

            describe('given the user made the necessary corrections', () => {

                beforeEach(() => {
                    handleUpdate.mockImplementation(() => { });
                    wrapper.find('#duration').simulate('change', {
                        target: {
                            value: "-5000",
                            name: "duration"
                        }
                    });
                });

                it("removes the old error's information", () => {
                    expect(wrapper.find('#durationLabel').prop('error')).toEqual(false);
                });
            });
        });
    });

});