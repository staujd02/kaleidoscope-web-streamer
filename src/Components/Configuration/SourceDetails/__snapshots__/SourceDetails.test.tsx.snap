// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Source Detail Component renders correctly 1`] = `
"
<div class=\\"MuiPaper-root MuiPaper-elevation1 MuiCard-root card source-detail MuiPaper-rounded\\">
  <div class=\\"MuiCardContent-root\\">
    <p class=\\"MuiTypography-root title MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom\\">Stream Details</p>
    <div class=\\"MuiFormGroup-root\\">
      <div class=\\"MuiFormControl-root MuiTextField-root\\">
        <label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled\\" data-shrink=\\"true\\">Source Name</label>
        <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\">
          <input type=\\"text\\" aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input\\" name=\\"title\\" value=\\"The Best Source\\">
        </div>
      </div>
      <div class=\\"MuiFormControl-root MuiTextField-root beefy\\">
        <label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled\\" data-shrink=\\"true\\">URL</label>
        <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\">
          <input type=\\"url\\" aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input\\" name=\\"source\\" value=\\"www.somewhere\\">
        </div>
        <p class=\\"MuiFormHelperText-root MuiFormHelperText-filled\\">Must be embeddable</p>
      </div>
      <div class=\\"MuiGrid-root\\">
        <label class=\\"MuiFormLabel-root MuiInputLabel-root inline-lb MuiInputLabel-animated\\" id=\\"durationLabel\\">Duration</label>
        <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline\\">
          <input type=\\"number\\" class=\\"MuiInputBase-input MuiInput-input\\" id=\\"duration\\" name=\\"duration\\" value=\\"6000\\">
        </div>
      </div>
      <div class=\\"MuiGrid-root\\">
        <label class=\\"MuiFormLabel-root MuiInputLabel-root inline-lb MuiInputLabel-animated\\">Enabled</label>
        <span class=\\"MuiSwitch-root\\">
          <span class=\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-239 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-240 Mui-checked\\" aria-disabled=\\"false\\">
            <span class=\\"MuiIconButton-label\\">
              <span class=\\"MuiSwitch-thumb\\"></span>
              <input type=\\"button\\" checked=\\"\\" class=\\"PrivateSwitchBase-input-242 MuiSwitch-input\\" name=\\"isEnabled\\">
            </span>
          </span>
          <span class=\\"MuiSwitch-track\\"></span>
        </span>
      </div>
    </div>
  </div>
</div>
"
`;
